content/blog/2026-09-01-atomic-slot-migration-acl-requirements/index.md

```
@@ -0,0 +1,215 @@
+blog_type = ["Technical Deep Dive"]
+[extra]
+featured = true
+featured_image = "/blog/atomic-slot-migration-acl-requirements/images/hero.png"
++++
+
+
+Valkey 9.0 introduced Atomic Slot Migration (ASM), a new resharding mechanism that replaces the traditional key-by-key migration process with a replication-inspired approach. Instead of copying individual keys using [`MIGRATE`](https://valkey.io/commands/migrate/), the source node establishes a dedicated connection to the target node and streams the contents of the migrating hash slots, followed by incremental updates until both nodes are synchronized. Once synchronization completes, ownership of the slot is transferred atomically.
+
+Since ASM internally reuses much of the replication infrastructure, it is natural to assume that it can reuse the same ACL permissions as a replica (which includes permission to execute the command [`CLUSTER SYNCSLOTS`](https://valkey.io/commands/cluster-syncslots/) also introduced in Valkey 9.0):
+
+```
```

markdownlint-cli2 (0.23.2)

[warning] Fenced code blocks should have a language specified

(MD040, fenced-code-language)